File: backend/services/predictor.py
"""Attendance bunk predictor — the math from CLAUDE.md §6.

Given classes attended (a) and conducted (c) and a target percentage T:

- can_skip:   most classes you can still miss and stay >= T.
              largest x with a / (c + x) >= T/100  ->  x = floor(a/(T/100) - c).
- must_attend: if already below T, fewest classes you must attend (in a row) to
              reach T.  smallest y with (a + y)/(c + y) >= T/100
              ->  y = ceil((T/100 * c - a) / (1 - T/100)).
"""
from __future__ import annotations

import math
from dataclasses import dataclass


@dataclass
class Prediction:
    percentage: float
    can_skip: int
    must_attend: int
    is_safe: bool


def predict(attended: int, conducted: int, threshold: float = 75.0) -> Prediction:
    """Compute attendance % and the bunk/attend numbers for one subject."""
    if conducted <= 0:
        return Prediction(percentage=0.0, can_skip=0, must_attend=0, is_safe=True)

    pct = attended / conducted * 100
    t = threshold / 100
    is_safe = pct >= threshold

    if is_safe:
        can_skip = max(0, math.floor(attended / t - conducted)) if t > 0 else 0
        must_attend = 0
    else:
        can_skip = 0
        # (1 - t) == 0 only if threshold is 100%, where you can never recover a miss.
        must_attend = (
            math.ceil((t * conducted - attended) / (1 - t)) if t < 1 else -1
        )
        must_attend = max(0, must_attend)

    return Prediction(
        percentage=round(pct, 2),
        can_skip=can_skip,
        must_attend=must_attend,
        is_safe=is_safe,
    )

File: frontend/src/lib/predictor.ts
// Client-side attendance predictor — mirrors backend/services/predictor.py so
// the target threshold and "what-if skip" projections recompute instantly from
// the raw attended/conducted counts, no refetch needed.

export type Prediction = {
  percentage: number;
  canSkip: number; // classes you can still miss and stay >= target
  mustAttend: number; // classes to attend in a row to reach target (0 if safe)
  isSafe: boolean;
};

export function predict(
  attended: number,
  conducted: number,
  threshold: number,
): Prediction {
  if (conducted <= 0) {
    return { percentage: 0, canSkip: 0, mustAttend: 0, isSafe: true };
  }
  const pct = (attended / conducted) * 100;
  const t = threshold / 100;
  const isSafe = pct >= threshold;

  // Epsilon guards: floating-point error can push an exact 7 to 7.0000002 and
  // flip floor/ceil by one (e.g. mustAttend showing 8 instead of 7).
  const EPS = 1e-9;
  let canSkip = 0;
  let mustAttend = 0;
  if (isSafe) {
    canSkip = t > 0 ? Math.max(0, Math.floor(attended / t - conducted + EPS)) : 0;
  } else {
    mustAttend =
      t < 1 ? Math.max(0, Math.ceil((t * conducted - attended) / (1 - t) - EPS)) : 0;
  }
  return { percentage: Math.round(pct * 100) / 100, canSkip, mustAttend, isSafe };
}

/** Projected % if you skip `skip` more classes (attend none of them). */
export function projectSkip(
  attended: number,
  conducted: number,
  skip: number,
): number {
  const c = conducted + skip;
  return c > 0 ? Math.round((attended / c) * 10000) / 100 : 0;
}

/** Projected % if you attend all of the next `attend` classes. */
export function projectAttend(
  attended: number,
  conducted: number,
  attend: number,
): number {
  const c = conducted + attend;
  return c > 0 ? Math.round(((attended + attend) / c) * 10000) / 100 : 0;
}
